runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -303,14 +309,34 @@ public void processElements(Iterable<WindowedValue<InputT>> values) throws Excep
       windowsToConsider.addAll(processElement(windowToMergeResult, value));
     }
 
-    // Trigger output from any window for which the trigger is ready
-    for (W mergedWindow : windowsToConsider) {
-      ReduceFn<K, InputT, OutputT, W>.Context directContext =
-          contextFactory.base(mergedWindow, StateStyle.DIRECT);
-      ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-          contextFactory.base(mergedWindow, StateStyle.RENAMED);
-      triggerRunner.prefetchShouldFire(mergedWindow, directContext.state());
-      emitIfAppropriate(directContext, renamedContext);
+    if (!windowsToConsider.isEmpty()) {
```

Comment:
I've refactored things some to make the flow more straightforward.
I moved some of the window stuff from processElement to improve their prefetching and also avoid needing to build up the consideration set upon what processElement returns.
I left some of the loops in the function since they were just a single line and a function for them seemed verbose, let me know if you'd still like them separated.